src/util/ssl_utils.js

```diff
@@ -34,105 +34,94 @@ function verify_ssl_certificate(certificate) {
 }
 
-// Get SSL certificates from local memroy.
-async function get_ssl_certificates() {
-    if (certs !== null) {
-        return certs;
+// Get SSL certificate (load once then serve from cache)
+function get_ssl_certificate(service) {
+    const cert_info = certs[service];
+    if (!cert_info) {
+        throw new Error(`Invalid service name, got: ${service}`);
+    }
+
+    if (cert_info.is_loaded) {
+        return cert_info.cert;
     }
 
-    return certs_read_sem.surround(async () => {
-        if (certs !== null) {
-            return certs;
+    return cert_info.sem.surround(async () => {
+        if (cert_info.is_loaded) {
+            return cert_info.cert;
         }
 
         try {
-            certs = await _read_ssl_certificates();
-            using_generated_certs = false;
+            cert_info.cert = await _read_ssl_certificate(cert_info.dir);
+            cert_info.is_generated = false;
             dbg.log0('Using mounted certificates');
 
         } catch (err) {
             if (err.code !== 'ENOENT') {
-                dbg.error('One or more SSL certificates failed to load', err.message);
-                dbg.warn('Fallback to generating self-signed certificates...');
+                dbg.error(`SSL certificate at ${cert_info.dir} failed to load`, err.message);
+                dbg.warn(`Fallback to generating self-signed for ${service}`);
             }
 
-            dbg.warn('Generating self-signed certificates');
-            certs = Object.keys(SSL_CERTS_DIR_PATHS)
-                .reduce((c, service_name) => {
-                        c[service_name] = generate_ssl_certificate();
-                        return c;
-                }, {});
-            using_generated_certs = true;
+            dbg.warn(`Generating self-signed certificates for ${service}`);
+            cert_info.cert = generate_ssl_certificate();
+            cert_info.is_generated = true;
         }
 
-        return certs;
+        cert_info.is_loaded = true;
+        return cert_info.cert;
     });
 }
 
-async function get_ssl_certificate(service) {
-    const all_certificates = await get_ssl_certificates();
-    const certificate = all_certificates[service] || null;
-    if (certificate === null) {
-        throw new Error(`Invalid service name, got: ${service}`);
-    }
-
-    return certificate;
-}
-
-function update_certs_from_disk() {
-    return certs_read_sem.surround(async () => {
-        try {
-            const certs_on_disk = await _read_ssl_certificates();
-            if (_compare_with_loaded_certs(certs_on_disk)) {
+// For each cert that was loaded into memory we check if the cert was changed on disk.
+// If so we update it. If any of the certs was updated we return true else we return false.
+async function update_certs_from_disk() {
+    const promiseList = Object.values(certs).map(cert_info =>
+        cert_info.sem.surround(async () => {
+            if (!cert_info.is_loaded || cert_info.is_generated) {
```

Comment:
if `is_generated` why are we skipping? shouldn't we load from disk and check in this case too?